design/httpproxy-prefix-rewrite.md

```diff
@@ -0,0 +1,262 @@
+This implies that even if the owner of a leaf HTTPProxy doesn't control the root or intermediate documents, they still have fairly concrete expectations around the full URL path to expect.
+
+Although there are a number of ways to conceptualize path prefix rewriting, path regex rewriting requires knowledge of the full request path at the time of writing the regex.
+This is because a regex rewrite can be used to re-order arbitratry portions of the URL path, and this can't be done unless the regex is defined as being applied to the complete path.
+
+## High Level Design
+
+Add a `PathRewrite` field to the HTTPProxy [Route](https://github.com/projectcontour/contour/blob/master/apis/projectcontour/v1/httpproxy.go#L125) object.
+This allows us to specify different kinds of rewrite for URL paths without also needing to specify a specific kind of request rewriting (i.e. we can add host rewrite later).
+
+Locating the `PathRewritePolicy` on the `Route` means that rewriting is controlled by the team that controls the leaf HTTPProxy document.
+This allows a useful initial implementation, but does not prevent adding `PathRewritePolicy` fields higher up the HTTPProxy tree (though allowing that would require detecting and resolving conflicts).
+
+```Go
+type Route struct 
+	...
+
+	// The policy for rewriting the path of the request URL
+	// after the request has been routed to a Service.
+	PathRewrite *PathRewritePolicy `json:"pathRewrite,omitempty"`
+}
+
+```
+
+The `PathRewritePolicy` struct initially only supports path prefix replacement.
+Creating a policy struct allows future addition of other kinds of path rewriting.
+
+
+```Go
+// PathRewritePolicy specifies how a request URL path should be
+// rewritten. This rewriting takes place after a request is routed
+// and has no subsequent effects on the proxy's routing decision.
+// 
+// Exactly one field in this struct may be specified.
+type PathRewritePolicy struct {
+    ReplacePrefix []struct {
+        // +optional
+        Prefix string `json:"prefix,omitempty"`
+
+        Replacement string `json:"replacement"`
+    } `json:"replacePrefix,omitempty"`
+}
+```
+
+`PrefixRewrite` is a slice to allow a service to rewrite multiple path prefixes, depending on the shape of the HTTPProxy tree.
+As discussed above, a HTTPProxy leaf can be reached from multiple HTTPProxy roots via different paths.
+This implies that service owners could reasonably need to rewrite multiple routing paths to a canonical version.
+This, in turn, implies that more than one prefix may need to be rewritten.
+
+For example, assume that a service support multiple versions of an API and is using the routing prefixes `/v1/api` and `/v2/api`.
+This can only be rewritten to `/api/v1` and `/api/v2` with two separate replacements.
+Distinguishing between these replacements requires specifying the routing prefix.
+
+* Prefix: This is the optional path prefix to match on.
```

Comment:
Is it possible to eliminate this feature from the first round of this design? It feels like it could be added later if required.

Reply:
Yes, we can ship without the `Prefix` field. In that case, the semantics will be that all prefixes will be replaced with the replacement. We should retain the enclosing struct so that we can gracefully add this later if necessary.